torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -353,6 +353,105 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
   return {std::move(payload), std::move(tensors)};
 }
 
+TensorPipeEntry tensorpipeSerialize(const Message& rpcMessage) {
+  tensorpipe::Message tpMessage;
+  std::vector<torch::Tensor> reservedTensors;
+  std::vector<std::vector<uint8_t>> copiedTensors;
+
+  const std::vector<char>& payload = rpcMessage.payload();
+  c10::List<at::Tensor> tensors = cloneSparseTensors(rpcMessage.tensors());
+
+  // Payload
+  tpMessage.data = (uint8_t*)(payload.data());
+  tpMessage.length = payload.size();
+
+  // Metadata - encode rpc message type + message id
+  tpMessage.metadata = c10::str(
+      static_cast<int>(rpcMessage.type()),
+      ":",
+      static_cast<int>(rpcMessage.id()));
+
+  // Tensors
+  tpMessage.tensors.reserve(tensors.size());
+  for (const at::Tensor& tensor : tensors) {
+    // Keep original user tensors and cloned sparse tensors
+    reservedTensors.push_back(tensor);
+    tensorpipe::Message::Tensor tpTensor;
+    torch::jit::Pickler pickler([&](const void* buf, size_t sz) -> size_t {
+      tpTensor.metadata.append(static_cast<const char*>(buf), sz);
+      return sz;
+    });
+    pickler.protocol();
+    pickler.pushIValue(tensor);
+    pickler.stop();
+    const auto& tensorDataVec = pickler.tensorData();
+    TORCH_INTERNAL_ASSERT(
+        tensorDataVec.size() == 1, "There should be single pickled tensor");
+    const auto& tensorData = tensorDataVec.front();
+    // Enforce memory copy if tensor is created from torch::from_blob, means
+    // that the tensor doesn't own the memory.
+    if (!tensorData.storageHasDeleter()) {
+      std::vector<uint8_t> storageData(tensorData.sizeInBytes());
+      memcpy(storageData.data(), tensorData.data(), storageData.size());
+      tpTensor.data = storageData.data();
+      copiedTensors.push_back(std::move(storageData));
+    } else {
+      tpTensor.data = (uint8_t*)(tensorData.data());
+    }
+    tpTensor.length = tensorData.sizeInBytes();
+    tpMessage.tensors.push_back(std::move(tpTensor));
+  }
+
+  return TensorPipeEntry{std::move(tpMessage),
+                         std::move(reservedTensors),
+                         std::move(copiedTensors)};
+}
+
+Message tensorpipeAllocateMessage(const tensorpipe::Message& tpMessage) {
+  // Payload, message type and message id
+  std::vector<char> payload(tpMessage.length);
+  size_t delimiterPos = tpMessage.metadata.find(":");
+  TORCH_INTERNAL_ASSERT(
+      delimiterPos != std::string::npos,
+      "message metadata must be delimitted by \":\" `",
+      tpMessage.metadata,
+      "`");
+  int mtypei = std::stoi(tpMessage.metadata.substr(0, delimiterPos));
+  uint64_t mId = std::stoull(tpMessage.metadata.substr(delimiterPos + 1));
+  MessageType mtype = static_cast<MessageType>(mtypei);
```

Comment:
We can avoid all this messy string ops if we make the first 4 bytes the message type and next 4 bytes the message id.